[Components] Update FontSize control

ntsekouras opened the pull request:
This PR is part of Typography tool update: https://github.com/WordPress/gutenberg/issues/34345 and is focuses on the `FontSizePicker` control.

The purpose of this PR is to try to see what is missing and how we will move forward to match the current design and also maybe revisit them to take into account our actual needs. We should make this iteratively and decide on these iterations. Also don't mind the actual code too much, as is mostly exploratory and will need refactoring. 

## Screenshots
### Current design
<img width="279" alt="Screen Shot 2021-09-24 at 2 00 42 PM" src="https://user-images.githubusercontent.com/25516611/134670861-ee2736d9-17a8-466c-8c49-b92b7725b88e.png">


| Predef (< 6 sizes) | Predef (>= 6 sizes) | Custom font size |
|---|---|---|
| ![image](https://user-images.githubusercontent.com/548849/131104825-5a3616bd-1b08-4ad4-bbe6-627791e1e7ee.png) | ![image](https://user-images.githubusercontent.com/548849/131104872-a61430ba-ce2b-4bd4-b7ea-f4b8045ad2f8.png) | ![image](https://user-images.githubusercontent.com/1083581/131821448-c1f3d0be-843a-456d-926a-aac3dd7b6936.png) |


And also from this: https://github.com/WordPress/gutenberg/issues/34345#issuecomment-926577197

<img width="820" alt="Screenshot 2021-10-06 at 1 10 50 PM" src="https://user-images.githubusercontent.com/16275880/136183755-48071b44-dc3a-41da-a6d1-f6d15f3fc794.png">

These first explorations have already raised some issues and some of them are things from changes  I've not committed here.

For a first iteration we could probably skip the dual control, but we'll see 😄  :
>When picking a font size from a list of predefined sizes, show either a ToggleGroup control (< 6 sizes) or a dropdown control (>= 6 sizes) (see screenshots below and [this early WIP prototype](https://g2-git-try-global-styles-sidebar-v2-itsjonq.vercel.app/iframe.html?id=examples-wip-fontsizepickerv2--default&viewMode=story))

## Notes

#### Control Label
We need to decide what should be shown next to the `Size` label in all cases (custom, from existing option, default, etc..)
<img width="280" alt="Screenshot 2021-10-06 at 1 15 51 PM" src="https://user-images.githubusercontent.com/16275880/136184423-6a97306a-5de1-46ee-b790-087c0a998f8c.png">

#### Custom Select component
Does changes in this components sound good like I've done in this PR? It would need better css handling though if we keep these..
Also the design is a bit different with the mock up as there is no space above the showed options.

#### Range Control
Currently the control supports only [unitless values](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/src/range-control/index.js#L273). We need to make this work with `units`.  We would also need 'position handling' support as it currently shows the `slider` on the left.
 


## Testing instructions
1. Just add a `paragraph` and check the `font size` control.

ntsekouras committed:
temp

Mamaduka commented:
Hi, Nik

I just wanted to note that we want this component to be backward compatible with unitless values. See #33679.

We have unit tests in place, so the regression should be easy to catch.

jasmussen commented:
Thanks for starting work on this one. It's a big one, it's a difficult one, and it's so nice to get something on the blank canvas. Here's a GIF of what I saw in a quick testing:

![font](https://user-images.githubusercontent.com/1204802/136201806-6c888a7f-e8d8-4c2f-87da-fb1073b022df.gif)

I'll need to come back to this one with deeper thoughts, but a few quick superficial observations:

- Would it make sense to make progress on the generic input-field component work? The 40px input fields from #34574 and #27331 might benefit the work here, just to have in place.
- The goal of the toolspanel is, in part, to retire the many Reset buttons strewn across the inspector. I wonder if we should tie this PR to #33744 by @aaronrobertshaw?
- The "Custom" option in the dropdown didn't seem to work for me, but hitting the settings icon did.

This is just one small ingredient of a large effort, nice work, and I think many of the pieces will start falling into place once we get this one down.

ntsekouras commented:
Thanks for the first tests @jasmussen and great remarks!

>Would it make sense to make progress on the generic input-field component work?
>The goal of the toolspanel is, in part, to retire the many Reset buttons strewn across the inspector. I wonder if we should tie this PR to Typography: Switch to ToolsPanel for display UI #33744 by @aaronrobertshaw?

Yes, it makes sense to fix/augment existing components that are going to be used. I made the draft to pause a bit and check some related PRs like the `ToolsPanel` that handle the `reset`, etc..

>The "Custom" option in the dropdown didn't seem to work for me, but hitting the settings icon did.

I know about that and haven't decided yet how to handle the options (default + custom) as it will also be coupled with `ToggleGroupControl`.

aaronrobertshaw commented:
> The goal of the toolspanel is, in part, to retire the many Reset buttons strewn across the inspector. I wonder if we should tie this PR to Typography: Switch to ToolsPanel for display UI #33744 by @aaronrobertshaw?

Thanks for the ping. It's great to see progress on the new typography components. 👍 

I did have some tweaks to the current `FontSizePicker` included in #33744 to allow the omission of the reset button and fix up spacing when used within the `ToolsPanel`. Those tweaks have just been [removed from that PR](https://github.com/WordPress/gutenberg/pull/33744/commits/de33264c66359f122b9561a8e3551a7126fd4404) although I planned to create a separate PR with them tomorrow.

If you'd prefer for me to put those changes on ice until this PR evolves, let me know. Even as it is now, the version of `FontSizePicker` in this PR would fit a little nicer within the `ToolsPanel` than the version on `trunk`. 

Other than the reset button, the only other tweak for the `ToolsPanel` would be removing the bottom margin so that the panel's grid can handle the spacing. 

Applying these changes alongside #33744 appears to work well and I'd expect it to evolve without many conflicts.

https://user-images.githubusercontent.com/60436221/136340922-4cdc2e0c-216d-4f07-b46e-ca9c96f3157e.mp4

I did encounter one small issue while testing which was in a couple of themes the select options included `NaN` in them. I believe this was due to the theme's font size presets using CSS vars for the size property. The Q theme is one [example](https://github.com/aristath/q/blob/master/theme.json#L10-L16).

<img width="286" alt="Screen Shot 2021-10-07 at 5 20 10 pm" src="https://user-images.githubusercontent.com/60436221/136340885-f69fc6d5-4223-4de8-ac77-7f39556a3162.png">

ciampo commented:
> Would it make sense to make progress on the generic input-field component work? The 40px input fields from The Global Styles Interface #34574 and Sidebar Controls & Component System #27331 might benefit the work here, just to have in place.

Hey @jasmussen , that definitely makes sense (although in a separate issue/PR). We should be adding any changes that are necessary to improve the typography panel in tracking issue #34345, so that we can take it from there and coordinate around these tasks.

With respect to what @aaronrobertshaw said, that sounds like a good plan to me!

ntsekouras commented:
>I did encounter one small issue while testing which was in a couple of themes the select options included NaN in them. I believe this was due to the theme's font size presets using CSS vars for the size property.

I'll look into this, thanks!

Regarding the reset I'm going to try to maybe have a first iteration with the reset button and the current `custom size` input with the units (no slider). We'll see how this goes and has to do with the other tools panel PRs. Remove the `reset` button will be easy 😄 .

ntsekouras committed:
show custom control on `custom` select option + show hints only if is simple css value

ntsekouras committed:
switch conditionally to ToggleGroupControl

ntsekouras committed:
add story part 1

ntsekouras commented:
I've updated the control to switch to `ToggleGroupControl` [conditionally](https://github.com/WordPress/gutenberg/pull/35395/files#diff-3b85bc4bfdaea5cd9cb6173bf9f52656ce4f656e5f846442fd15185a5b78fb1dR51) and added a simple story for start to test there as well(`Different Control By Size` under `FontSizePicker`).

I think it's ready for another round of testing with different themes etc 😄

ntsekouras commented on packages/components/src/custom-select-control/style.scss:
```diff
@@ -58,19 +58,25 @@
 
-
+	&.has-hint {
+		grid-template-columns: auto auto 30px;
```
Can we handle this better? 🤔 --cc @jasmussen

ciampo replied:
Have you considered using the [`Grid` component](https://github.com/WordPress/gutenberg/tree/trunk/packages/components/src/grid) instead of using CSS styles?

jasmussen replied:
Can you expand on what I should look for here? Happy to help if I can.

ciampo replied:
What I mean was to refactor this component to use internally the `Grid` component — instead of using custom CSS rules for the grid layout.

ntsekouras replied:
I wanted to make the minimum changes in CustomSelectControl and I'm not sure there are some better rules I should use - or even maybe `flex`? Maybe I'll consider splitting these changes to a separate PR as Marco suggested..

jasmussen commented:
Took this one for a quick spin, and my goodness, even though parts are missing, this is just a big step forward:
![typography](https://user-images.githubusercontent.com/1204802/136559294-deab1c3e-ba64-4081-b010-d722b768b1d2.gif)

In the mockups, the fields use a new all-caps subheading style. I imagine this is something we can follow up with separately, but for now we should at least remove the thin font weight from this one:
<img width="110" alt="Screenshot 2021-10-08 at 14 44 00" src="https://user-images.githubusercontent.com/1204802/136559543-cf3c7968-8ad0-4910-a70d-e6b56fd18526.png">

Instead of `font-weigh: 200;` it should just be normal, and you can then set it to `color: $gray-700;` and perhaps give it `$grid-unit-05` left margin.

mtias committed:
Update comments.

ciampo commented on packages/components/src/toggle-group-control/toggle-group-control-option.tsx:
```diff
@@ -56,7 +61,9 @@ function ToggleGroupControlOption(
-	const showSeparator = getShowSeparator( toggleGroupControlContext, index );
+	const hasItems = !! toggleGroupControlContext.items.length;
+	const showSeparator =
+		hasItems && getShowSeparator( toggleGroupControlContext, index );
```
With #35497 merged, we don't need these changes anymore

ciampo commented on packages/components/src/toggle-group-control/toggle-group-control-backdrop.tsx:
```diff
@@ -29,7 +29,11 @@ function ToggleGroupControlBackdrop( {
 			`[data-value="${ state }"]`
 		);
-		if ( ! targetNode ) return;
+		if ( ! targetNode ) {
```
In which scenario would `targetNode` not exist? Should we avoid rendering the `BackdropView` in that case?

ntsekouras replied:
Actually this is my fix for issue: 35383. A contributor has expressed some interest and have let them know is a bit time sensitive and offered to help if they need to.

The thing is that this is probably related to `Reakit` and was related to the fix here: https://github.com/WordPress/gutenberg/pull/35409 (that got removed now). The problem is probably that if we have a value in Reakit `state` that doesn't exist in our options, the `currentId` is filled with the first available option, which is not what we want. [Reakit docs](https://reakit.io/docs/radio/#useradiostate) state that if we pass `null` the `currentId` will not be set, but with some tests of mine, it still had the first option as value.

This was mentioned [here by you](https://github.com/WordPress/gutenberg/pull/31937#discussion_r666004949) and there was not a definite answer found, even by @diegohaz - so I guess it would be the above? 🤔 

If this is to be addressed it should be in a separate PR and it might be a bit trickier as we should handle `undefined` values and change them to `null`?

ciampo replied:
Definitely better to address it separately. It will allow us to focus on testing for that issue specifically, write better docs/changelog and potentially tests. It will be also easier for the contributor interested in the fix to understand it.

> This was mentioned here by you and there was not a definite answer found,

Ha! Well spotted. I had a feeling that we were going to meet with that line of code soon enough 😆 

> it might be a bit trickier as we should handle `undefined` values and change them to `null`?

I still don't fully understand what's the better solution here. It would definitely be great if @diegohaz could help us with that

ciampo replied:
Closing this conversation, since #35545 was opened to deal with this separately

ciampo replied:
Closing this conversation, since #35545 was opened to deal with this separately

ciampo commented on packages/components/src/font-size-picker/index.js:
```diff
@@ -60,104 +34,207 @@ function FontSizePicker(
-			className="components-font-size-picker"
-			{ ...( ref ? {} : { ref } ) }
-		>
+		<fieldset className={ baseClassName } { ...( ref ? {} : { ref } ) }>
```
The `fieldset` wrapper has been removed from this component in #35461

ntsekouras replied:
It's been removed from `FontAppearanceControl` - this is `FontSizePicker`. It got me confused at first if I made some conflicts resolution wrong 😄

ciampo commented on packages/components/src/font-size-picker/index.js:
```diff
@@ -60,104 +34,207 @@ function FontSizePicker(
+		selectedOption.name
 	);
-
+	const baseClassName = 'components-font-size-picker';
```
I know that this classname existed before this PR, but it got me thinking — it'd be also great if we could ride the wave of changes to `FontSizePicker` to also switch its styling from Sass to Emotion, and convert it to TypeScipt (of course in a separate PR)

ntsekouras replied:
Sounds good, but let's do this separately..

ciampo replied:
Yep, that sounds like a good follow-up

ciampo commented on packages/components/src/font-size-picker/index.js:
```diff
@@ -60,104 +34,207 @@ function FontSizePicker(
+	//
+	// A select control is also used when the value of a preset cannot be
+	// immediately computed (eg. 'calc', 'var').
+	const useSelectControl =
```
Bit of a nit, but usually in React components variables beginning with `use...` represent react hooks. Maybe we could rename it to `shouldUseSelectControl` ?

ntsekouras replied:
Yeah, I thought about this and meant to change it, but then forgot about it 😄

ciampo commented on packages/components/src/custom-select-control/index.js:
```diff
@@ -98,14 +98,9 @@ export default function CustomSelectControl( {
-			0,
-			'downshift-null'.length
-		) === 'downshift-null'
+		menuProps[ 'aria-activedescendant' ]?.startsWith( 'downshift-null' )
```
Nice refactor!

ciampo commented on packages/components/src/custom-select-control/index.js:
```diff
@@ -161,12 +156,18 @@ export default function CustomSelectControl( {
 							} ) }
 						>
 							{ item.name }
+							{ item.hint && (
+								<span className="components-custom-select-control__item-hint">
+									{ item.hint }
+								</span>
+							) }
```
I think it would be better to split changes to `CustomSelectControl` in a separate PR, where the relevant documentation is also updated and a CHANGELOG entry is added.

The current PR will keep focusing on the `FontSizePicker`, making changes more contained and easier to find in the future.

---

Separately from this, I also wonder if we should rewrite the `CustomSelectControl` component to use Emotion for styling and rewrite it to TypeScript. These improvements would ease the integration of these components with the Typography panel and give more confidence to our refactors (via TYpeScript's static linting)

ciampo reviewed (commented):
Thank you for starting work on this PR, @ntsekouras ! As others have also said, it definitely represents a big improvement over the current UI.

I've been taking a more deeper look at the code and left a few comments. We should also rebase this PR to include a few changes to the `FontSizePicker` component that got merged in the meantime.

Regarding the proposal of rewriting these components to TypeScript and use Emotion, I'm more than happy to help (either directly, or indirectly with advice and reviews).

Finally, I just wanted to cross-post the fact that #35451 may contain some overlapping changes (cc'ing the PR author, @aaronrobertshaw )

ntsekouras committed:
fix styles + rename var

ntsekouras commented:
>In the mockups, the fields use a new all-caps subheading style. I imagine this is something we can follow up with separately,

Yes. I saw that but this affects all control labels and we should be consistent. In a separate PR we should update all controls to uppercase - css probably(?) as third party controls can be there..

>we should at least remove the thin font weight from this one:

Good call 😄 - I updated the styles.

ciampo commented:
> Yes. I saw that but this affects all control labels and we should be consistent. In a separate PR we should update all controls to uppercase - css probably(?) as third party controls can be there..

I believe a similar issue was raised in https://github.com/WordPress/gutenberg/pull/35400#discussion_r724311248. We've got #35464 open to potentially update the `Heading` component styles to match the latest design requirements.

aaronrobertshaw commented:
> Finally, I just wanted to cross-post the fact that #35451 may contain some overlapping changes (cc'ing the PR author, @aaronrobertshaw )

The changes in #35451 are only to clean up the display of the existing versions of typography controls when displayed within the `ToolsPanel`. This was when the plan, as I understood it, was for that typography panel switch to occur before the new typography components landed.

Given the rapid progress here (great work @ntsekouras!) I don't think it makes a lot of sense to introduce the changes in #35451 if they may conflict with or distract from this work. I'm more than happy to keep monitoring all these changes and complete the switching of the typography panel to use `ToolsPanel` once this lands.

ntsekouras commented:
>I don't think it makes a lot of sense to introduce the changes in #35451 if they may conflict with or distract from this work.

Thanks Aaron! I'm planning to review the tools panel PRs and we'll see how we move forward. We can handle the conflicts either way 😄

ciampo commented:
Aaron's plan sounds good to me. I've closed #35451, we can continue development of the `FontSizePicker` here for the time being!

ntsekouras committed:
update README

ntsekouras committed:
small fixes and update unit tests

ntsekouras commented on packages/components/src/font-size-picker/index.js:
```diff
@@ -60,104 +34,208 @@ function FontSizePicker(
+					<FlexItem>
+						<Button
+							label={ __( 'Toggle custom size display' ) }
+							showTooltip={ false }
```
What `label` should we add here?

mtias replied:
"Set custom size" and "Use size preset"?

ntsekouras commented on packages/components/src/font-size-picker/test/index.js:
```diff
@@ -84,6 +91,7 @@ describe( 'FontSizePicker', () => {
 				/>
 			);
 
+			toggleCustomInput();
```
I'd love some feedback for these tests from @Mamaduka, if you find some time 😄 . I'll add e2e tests soon.

Mamaduka replied:
Thanks for the ping, @ntsekouras. I'll try to tests and provide feedback tomorrow.

Mamaduka replied:
I think changes look good here.

I also tested FontSize control updates with the same method I used for #33679, and it works as expected.

@ntsekouras, feel free to ping me once the PR is out of the draft state. I'm happy to do another round of tests.

ntsekouras replied:
Thanks so much George!

ntsekouras committed:
update e2e tests part 1

ntsekouras committed:
update current e2e tests + add aria label to toggle group button

ntsekouras committed:
update toggle custom size label

ntsekouras committed:
add tests + reorganise them

jasmussen commented:
Taking this one for a spin again, it feels pretty solid to me:
![font size](https://user-images.githubusercontent.com/1204802/137313717-f2655129-fee9-4866-9863-127ebed93e3e.gif)

There are a number of small things we need to fix, but I don't think they need to happen as part of this PR. For example, the custom font size icon is offset from the right:
<img width="347" alt="Screenshot 2021-10-14 at 14 05 56" src="https://user-images.githubusercontent.com/1204802/137314095-1ca03747-e6ce-487f-a321-41f1b2bf9dfa.png">

This padding is also what makes the toggle wider, and scales down the icon inside:

<img width="77" alt="Screenshot 2021-10-14 at 14 06 48" src="https://user-images.githubusercontent.com/1204802/137314210-20c4a16a-3115-46c5-b211-88cfdb6fcb2a.png">

The _right_ way to fix that is to revisit the CSS that adds this padding in the first place, which appears to be this:

```
.components-button.is-small.has-icon:not(.has-text) {
    padding: 0 8px;
    width: 24px;
}
```

However in the name of shipping this, we _could_ add the following in the mean time? Happy to push if need be. This would fix it:

```
.components-font-size-picker__header .components-button.is-small.has-icon:not(.has-text) {
    min-width: $icon-size;
    padding: 0;
}
```
<img width="295" alt="Screenshot 2021-10-14 at 14 10 23" src="https://user-images.githubusercontent.com/1204802/137314739-c66f41d2-1e11-4f0b-b12d-678462ec8039.png">
<img width="293" alt="Screenshot 2021-10-14 at 14 12 45" src="https://user-images.githubusercontent.com/1204802/137315067-6a195cb1-c737-43a7-ba7f-783f91a8bbe3.png">

jasmussen commented:
Also wanted to note that the items inside the segmented control follow the order of font sizes defined in theme.json. I had them in large to small descending order, and when correcting them to be small to large, this is reflected in the component:

```
			"fontSizes": [
				{
					"slug": "small",
					"size": "12px",
					"name": "Small"
				},
				{
					"slug": "large",
					"size": "18px",
					"name": "Large"
				},
				{
					"slug": "medium",
					"size": "14px",
					"name": "Medium"
				},
				{
					"slug": "xl",
					"size": "48px",
					"name": "XL"
				},
				{
					"slug": "xxl",
					"size": "64px",
					"name": "XXL"
				}
			],
```
<img width="298" alt="Screenshot 2021-10-14 at 14 15 19" src="https://user-images.githubusercontent.com/1204802/137315484-eef6adea-9180-4c4a-8f57-5858fafee186.png">

👍 👍

jasmussen commented:
I can also confirm that if I add additional values, more than 5, they show up in the custom dropdown:
<img width="319" alt="Screenshot 2021-10-14 at 14 17 03" src="https://user-images.githubusercontent.com/1204802/137315659-e1c85530-aa17-4b98-a00c-3297adc479b1.png">

Edit: Actually, let's replace that font-weight: 200; with a $gray-700 color.

And the dropdown looks reasonably good, and can be improved separately. From a design point of view, I'd be happy to greenlight this one if we can fix the custom font size icon, as [outlined here](https://github.com/WordPress/gutenberg/pull/35395#issuecomment-943296291), and I'd also personaly be okay with bespoke CSS to fix it, and then follow up on that separately.

jasmussen committed:
Polish icon size and position, polish font size in dropdown.

jasmussen commented:
Since there's already a CSS file that customizes this panel, I went ahead and pushed a small fix to the icon size, padding, position and width, and changed the font weight in the dropdown. Looks good to me now:
<img width="305" alt="Screenshot 2021-10-14 at 14 23 30" src="https://user-images.githubusercontent.com/1204802/137316720-3352d42c-48e0-4a30-9640-f8c992a0c83a.png">
<img width="323" alt="Screenshot 2021-10-14 at 14 23 39" src="https://user-images.githubusercontent.com/1204802/137316730-d3c07729-9840-4839-84da-4bc36f810d7b.png">

jasmussen commented:
Tested other instances of the CustomSelectControl, to verify they hadn't regressed. It's used for font appearance, and for date format. Easy ways to test is to insert a Post Date block and check the format in the inspector:
![post comment date](https://user-images.githubusercontent.com/1204802/137317356-1a33e511-84a9-4821-a500-e8de482e4c5d.gif)

Or the tagline block and check the font appearance:
![tagline](https://user-images.githubusercontent.com/1204802/137317390-4f82c507-30ab-4c92-9150-2bf927355a6e.gif)

All looks solid to me.

jasmussen reviewed (approved):
This feels solid to me, and a big step forward. From the design side, there are a few small things to follow up on, but most of them fall in the category of general components work (40px input fields, consume components without having to write bespoke CSS), and not something that should block this PR.

I tested it with a few font size presets, and with many. In both cases the experience works well for me, using both mouse and keyboard. I also verified that the enhancements made to the CustomSelectControl have not caused trouble with any of the existing usages of the component. In that light, this one feels safe.

I'd be happy to land this one as a first iteration and then follow up if anything appears.

geriux commented:
Hey @ntsekouras 👋 I just tested these changes on mobile and it's working well, thanks for the ping! 👍

ntsekouras commented:
Thank you all for your help here!!! 🙏

jasmussen commented:
🔥

ciampo commented:
It's good to see progress happening so fast!

One thing that I'd like to note, though, is that we included changes to `CustomSelectControl` which [I believe should have been split to a separate PR](https://github.com/WordPress/gutenberg/pull/35395#discussion_r726484560) (could we do it in a follow-up PR at least?) and that we should try to avoid adding more hardcoded classnames to component's styles.

Introducing a lot of fast changes, especially in a package like the components package, can become counter-productive very quickly.

ciampo commented:
Updated `CustomSelectControl` docs in this follow-up PR #35673

ntsekouras commented:
>One thing that I'd like to note, though, is that we included changes to CustomSelectControl which I believe should have been split to a separate PR
>Introducing a lot of fast changes, especially in a package like the components package, can become counter-productive very quickly.

You're right about that and I was thinking maybe it's better to make this prop experimental for now and revisit when the typescript/css-in-js is handled for this component ? 🤔
